### ISSUE TITLE
Enum `Literal`s not accepted in methods

### ISSUE THREAD

**Issue by patrick-kidger** [microsoft/pyright#4502]
I'm looking to pass around types that are either an integer, or some sentinel value.

I think the best way to do this in a static-type-checking-compatible-way is to make the sentinel value be an enum with a single variant. (Since enum entries are allowed in `Literal`.)

This approach seems to work in free functions, but produce a false positive when used in methods. Here's a MWE:
```python
import enum
from typing import Literal

class Foo(enum.Enum):
    foo = enum.auto()

foo = Foo.foo
IntOrFoo = int | Literal[foo]

def accepts_int(x: int):
    ...

# This is fine
def call(maybe_int: IntOrFoo):
    if maybe_int is not foo:
        accepts_int(maybe_int)

class Call:
    def __init__(self, maybe_int: IntOrFoo):
        self.maybe_int = maybe_int

    # This is not fine
    def call(self):
        if self.maybe_int is not foo:
            accepts_int(self.maybe_int)
```

`pyright myfile.py` produces:

```
pyright 1.1.290
/some/path/to/myfile.py
  /some/path/to/myfile.py:25:25 - error: Argument of type "int | Foo" cannot be assigned to parameter "x" of type "int" in function "accepts_int"
    Type "int | Foo" cannot be assigned to type "int"
      "Foo" is incompatible with "int" (reportGeneralTypeIssues)
```
Expected behaviour is for no error to be raised.

**Comment by erictraut** [comment]
There are a couple of issues in your code. First, you are attempting to use the variable `foo` within a type annotation. Variables aren't generally allowed within type annotations. Is your intent for `foo` to be a type alias? If so, it should be defined as `foo: TypeAlias = Literal[Foo.foo]`.

Second, since `foo` is a variable without a type annotation, its type will be inferred. When it's accessed within the same execution context as its scope, pyright will infer its type based on the code flow, which takes into account type narrowing based on conditional expressions (type guards) and assignments. If the same variable is accessed from another scope, such as within a class, its type will be determined based on assigned values, but type narrowing will not apply. In your case, the inferred type of `foo` is `Foo`. The scope of this variable is the module, so type narrowing will apply when it is accessed within that scope. When it's accessed in a different execution context (e.g. within the `Call.call` method), type narrowing doesn't apply, and its type is assumed to be `Foo`. That explains the behavior you're seeing.

You could use define `foo` as `Final` to indicate that it's constant and cannot be reassigned some other value: `foo: Final = Foo.foo`. In this case, its type will be inferred as `Literal[Foo.foo]` rather than `Foo`.

**Comment by patrick-kidger** [comment]
Thanks for the help! That make sense. Fixing these issues, I think I've found another false positive (assuming I'm not missing something else):
```python
import enum
from typing import Final, Literal, TypeAlias

class Foo(enum.Enum):
    foo = enum.auto()

foo: Final = Foo.foo
IntOrFoo: TypeAlias = int | Literal[Foo.foo]

def call(maybe_int: IntOrFoo):
    ...

class Cls:
    def __init__(self, maybe_int: IntOrFoo):
        self.maybe_int = maybe_int

    def method(self):
        call(self.maybe_int)
```
raises:
```
pyright 1.1.290
/path/to/myfile.py
  /path/to/myfile.py:18:14 - error: Argument of type "int | Foo" cannot be assigned to parameter "maybe_int" of type "IntOrFoo" in function "call"
    Type "int | Foo" cannot be assigned to type "IntOrFoo"
      "Foo" cannot be assigned to type "Literal[Foo.foo]"
      Type "Foo" cannot be assigned to type "IntOrFoo"
        "Foo" is incompatible with "int"
        "Foo" cannot be assigned to type "Literal[Foo.foo]" (reportGeneralTypeIssues)
```
In this case the same `IntOrFoo` annotation is now being used throughout, it's just being interpretered in different ways in different contexts.

Adding a few `reveal_type`s, it seems like `reveal_type(self.maybe_int)` switches from `int | Literal[Foo.foo]` to `int | Foo` between the `__init__` and `method` calls.

**Comment by erictraut** [comment]
Pyright is working as designed. You have provided no type declaration for the instance variable `maybe_int`, so its type is inferred based on values that are assigned to it. In your sample, it is assigned the expression `maybe_int`. When pyright infers the type of an instance variable, it eliminates any literal values and widens them to their non-literal counterpart. This is generally the expected behavior. Consider the following:

```python
class Foo:
    def __init__(self):
        self.my_var = 1
```

In the above example, you probably expect the type of `my_var` to be `int` rather than `Literal[1]`.

If your intent is for `maybe_int` to retain its literal type, you must provide an explicit type declaration to that effect.

```python
        self.maybe_int: IntOrFoo = maybe_int
```

**Comment by patrick-kidger** [comment]
Ah, I didn't realise there was that interact between instance variables and literals.

Thank you for your help.
